### PR TITLE
[ADD] fsm_location_builder

### PR DESCRIPTION
**Pull request by osi-scampbell**
This PR adds fsm_location_builder, a tool used to create multiple fsm_locations in a short period of time. This way, when a user has a building with 200 units on 4 floors in 2 buildings, the user can use this tool instead of manually creating all 2000 units manually.

#128

### CODE REVIEW

File: fieldservice/models/res_config_settings.py [hunk 1]
```diff
@@ -53,6 +53,12 @@ class ResConfigSettings(models.TransientModel):
         string='Manage Recurring Orders')
     module_fieldservice_project = fields.Boolean(
         string='Projects and Tasks')
+    module_fieldservice_location_builder = fields.Boolean(
+        string='User FSM Location Builder')
+    auto_populate_persons_on_location = fields.Boolean(
+        string='Auto-populate Workers on Location based on Territory',
+        related='company_id.auto_populate_persons_on_location',
+        readonly=False)
```
Comment (max3903) [inline fieldservice/models/res_config_settings.py]: @osi-scampbell You already have that field below.

File: fieldservice/views/res_config_settings.xml [hunk 2]
```diff
@@ -254,6 +254,17 @@
...
+                            <div class="o_setting_left_pane">
+                                 <field name="module_fieldservice_location_builder"/>
+                            </div>
+                            <div class="o_setting_right_pane">
+                                <label for="fsm_location_builder" string="FSM Location Builder"/>
+                                <div class="text-muted">
+                                    Create multiple FSM Locations at once using Location Builder Wizard
+                                </div>
+                            </div>
+                        </div>
```
Comment (max3903) [inline fieldservice/views/res_config_settings.xml]: @osi-scampbell Should be in the "Locations" section

File: fieldservice_location_builder/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Copyright (C) 2019 - TODAY, Open Source Integrators
```
Comment (max3903) [inline fieldservice_location_builder/__init__.py]:
```suggestion
# Copyright (C) 2019 Open Source Integrators
```

File: fieldservice_location_builder/views/fsm_location_level.xml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                <field name="name"/>
+                <field name="spacer"/>
+                <field name="start_number"/>
+                <field name="end_number"/>
+                <field name="total_number"/>
+            </tree>
+        </field>
+    </record>
+
+    <!-- District Form View -->
```
Comment (max3903) [inline fieldservice_location_builder/views/fsm_location_level.xml]:
```suggestion
    <!-- Location Level Form View -->
```

File: fieldservice_location_builder/wizard/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# Copyright (C) 2019 - TODAY, Open Source Integrators
```
Comment (max3903) [inline fieldservice_location_builder/wizard/__init__.py]:
```suggestion
# Copyright (C) 2019 Open Source Integrators
```

File: fieldservice_location_builder/wizard/fsm_location_builder_wizard.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# Copyright (C) 2019 - TODAY, Open Source Integrators
```
Comment (max3903) [inline fieldservice_location_builder/wizard/fsm_location_builder_wizard.py]:
```suggestion
# Copyright (C) 2019 Open Source Integrators
```

**Comment by max3903** [comment]
@osi-scampbell You are missing the `readme` directory with the content from #128

**Review by patrickrwilson — CHANGES_REQUESTED** [review]
Nice work, I'd like to make some requests:

- [ ] Defaults: I think when a sub location is created it should have the following default values inherited from the primary location.

- Street 1
- Street 2 Shouldn't default from parent but maybe the name sub location name should go here?
- City
- State
- Zip
- Country
- Timezone
- Territory

- [ ] Tags: Can tags be added to the wizard? The workflow could be that when you add a line, say building, you can select tags that will be on those records (like a building tag), and on the next line, you can have Units with the unit tag, etc.

**Comment by max3903** [comment]
@osi-scampbell ping

**Comment by patrickrwilson** [comment]
I tried to create a typical sample sub locations like the following...
image

Note that I created the tags on the fly in the wizard view using the quick create.
After clicking the create button, I get the following error...
  File "C:\Program Files (x86)\Odoo 12.0e\server\odoo\api.py", line 736, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 57, in create_sub_locations
    build_location(location, 0)
  File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 56, in build_location
    build_location(new_location, num + 1)
  File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 56, in build_location
    build_location(new_location, num + 1)
  File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 27, in build_location
    tags = self.level_ids[lev_num].tag_ids.ids
  File "C:\Program Files (x86)\Odoo 12.0e\server\odoo\models.py", line 5121, in __getitem__
    return self._browse((self._ids[key],), self.env)
IndexError: tuple index out of range

**Comment by osi-scampbell** [comment]
> I tried to create a typical sample sub locations like the following...
> image
> 
> Note that I created the tags on the fly in the wizard view using the quick create.
> After clicking the create button, I get the following error...
> File "C:\Program Files (x86)\Odoo 12.0e\server\odoo\api.py", line 736, in _call_kw_multi
> result = method(recs, *args, **kwargs)
> File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 57, in create_sub_locations
> build_location(location, 0)
> File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 56, in build_location
> build_location(new_location, num + 1)
> File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 56, in build_location
> build_location(new_location, num + 1)
> File "c:\program files (x86)\odoo 12.0e\server\odoo\addons\fieldservice_location_builder\wizard\fsm_location_builder_wizard.py", line 27, in build_location
> tags = self.level_ids[lev_num].tag_ids.ids
> File "C:\Program Files (x86)\Odoo 12.0e\server\odoo\models.py", line 5121, in **getitem**
> return self._browse((self._ids[key],), self.env)
> IndexError: tuple index out of range

Error Solved, was using 1-3 for my building/floor/unit testing, jumping the number up to 100-105 revelead a variable that was out of place throwing that error. Made fix, tested using those numbers, locataions built. Let me know if it works on your end as well.

File: fieldservice_location_builder/views/fsm_location_level.xml [hunk 7]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (patrickrwilson) [inline fieldservice_location_builder/views/fsm_location_level.xml]: Not needed for 12 and above

**Review by patrickrwilson — APPROVED** [review]
Works great. Just a small suggestion and please fix the lint error. Good to go after that.

**Review by max3903 — APPROVED** [review]

**Pull request merged**